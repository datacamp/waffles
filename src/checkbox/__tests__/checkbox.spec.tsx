import { useRef, useEffect } from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Checkbox } from '../index';

function TestRefCheckbox() {
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    inputRef.current?.focus();
  }, []);

  return (
    <Checkbox ref={inputRef} onChange={jest.fn()}>
      Test label
    </Checkbox>
  );
}

describe('Checkbox', () => {
  it('renders label and input', () => {
    const { getByLabelText, getByText } = render(
      <Checkbox onChange={jest.fn()}>Taylor Swift</Checkbox>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });
    const label = getByText('Taylor Swift');

    expect(input).toBeInTheDocument();
    expect(label).toBeInTheDocument();
  });

  it('handles click event correctly', () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <Checkbox onChange={handleClick}>Taylor Swift</Checkbox>,
    );

    const label = getByText('Taylor Swift');
    fireEvent.click(label);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('handles focus event correctly', () => {
    const handleFocus = jest.fn();
    const { getByLabelText } = render(
      <Checkbox onFocus={handleFocus} onChange={jest.fn()}>
        Taylor Swift
      </Checkbox>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });
    fireEvent.focus(input);

    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it('sets the data attribute on the checkbox', () => {
    const { getByTestId } = render(
      <Checkbox data-testid="taylors-checkbox" onChange={jest.fn()}>
        Taylor Swift
      </Checkbox>,
    );

    const input = getByTestId('taylors-checkbox');

    expect(input).toBeInTheDocument();
  });

  it('sets correct aria attributes when has error', () => {
    const { getByLabelText } = render(
      <Checkbox error onChange={jest.fn()}>
        Taylor Swift
      </Checkbox>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });

    expect(input).toHaveAttribute('aria-invalid', 'true');
  });

  it('renders the disabled checkbox', () => {
    const { getByLabelText } = render(
      <Checkbox disabled onChange={jest.fn()}>
        Taylor Swift
      </Checkbox>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });

    expect(input).toBeDisabled();
  });

  it('accepts ref and could be focused programmatically', () => {
    const { container } = render(<TestRefCheckbox />);

    const input = container.querySelector('input');

    expect(input).toHaveFocus();
  });

  it('renders basic snapshot', () => {
    const { container } = render(
      <Checkbox onChange={jest.fn()}>Taylor Swift</Checkbox>,
    );

    const checkbox = container.firstChild;

    expect(checkbox).toMatchSnapshot();
  });

  it('renders snapshot of checked state', () => {
    const { container } = render(
      <Checkbox checked onChange={jest.fn()}>
        Taylor Swift
      </Checkbox>,
    );

    const checkbox = container.firstChild;

    expect(checkbox).toMatchSnapshot();
  });

  it('renders snapshot with error', () => {
    const { container } = render(
      <Checkbox error onChange={jest.fn()}>
        Taylor Swift
      </Checkbox>,
    );

    const checkbox = container.firstChild;

    expect(checkbox).toMatchSnapshot();
  });

  it('renders snapshot of inverted', () => {
    const { container } = render(
      <Checkbox inverted onChange={jest.fn()}>
        Taylor Swift
      </Checkbox>,
    );

    const checkbox = container.firstChild;

    expect(checkbox).toMatchSnapshot();
  });

  it('renders snapshot of disabled state', () => {
    const { container } = render(
      <Checkbox disabled onChange={jest.fn()}>
        Taylor Swift
      </Checkbox>,
    );

    const checkbox = container.firstChild;

    expect(checkbox).toMatchSnapshot();
  });

  it('renders snapshot of focused state', () => {
    const { container, getByLabelText } = render(
      <Checkbox onChange={jest.fn()}>Taylor Swift</Checkbox>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });
    fireEvent.focus(input);
    const checkbox = container.firstChild;

    expect(checkbox).toMatchSnapshot();
  });
});
