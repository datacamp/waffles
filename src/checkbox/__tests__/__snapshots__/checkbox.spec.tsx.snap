// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders basic snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 0;
  height: 0;
  margin: 0;
  padding: 0;
  overflow: hidden;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 18px;
  width: 18px;
  border-radius: 4px;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  color: #FFFFFF;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
}

<label
  class="emotion-0"
>
  <input
    aria-invalid="false"
    class="emotion-1"
    type="checkbox"
  />
  <div
    class="emotion-2"
  />
  <div
    class="emotion-3"
  >
    Taylor Swift
  </div>
</label>
`;

exports[`Checkbox renders snapshot of checked state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 0;
  height: 0;
  margin: 0;
  padding: 0;
  overflow: hidden;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 18px;
  width: 18px;
  border-radius: 4px;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  color: #FFFFFF;
  background-color: #009BD8;
  border-width: 1px;
  border-style: solid;
  border-color: #009BD8;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
}

<label
  class="emotion-0"
>
  <input
    aria-invalid="false"
    checked=""
    class="emotion-1"
    type="checkbox"
  />
  <div
    class="emotion-2"
  >
    <svg
      aria-hidden="true"
      height="14"
      viewBox="0 0 18 18"
      width="14"
    >
      <path
        d="M13.746 4.337a1.015 1.015 0 0 1 1.409-.099c.417.354.462.97.101 1.378l-7.13 8.047a1.015 1.015 0 0 1-1.483.03L2.771 9.67a.961.961 0 0 1 .044-1.38 1.015 1.015 0 0 1 1.412.041l3.113 3.235 6.406-7.229Z"
        fill="currentColor"
      />
    </svg>
  </div>
  <div
    class="emotion-3"
  >
    Taylor Swift
  </div>
</label>
`;

exports[`Checkbox renders snapshot of disabled state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: default;
  opacity: 0.6;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 0;
  height: 0;
  margin: 0;
  padding: 0;
  overflow: hidden;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 18px;
  width: 18px;
  border-radius: 4px;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  color: #FFFFFF;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
}

<label
  class="emotion-0"
>
  <input
    aria-invalid="false"
    class="emotion-1"
    disabled=""
    type="checkbox"
  />
  <div
    class="emotion-2"
  />
  <div
    class="emotion-3"
  >
    Taylor Swift
  </div>
</label>
`;

exports[`Checkbox renders snapshot of focused state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 0;
  height: 0;
  margin: 0;
  padding: 0;
  overflow: hidden;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 18px;
  width: 18px;
  border-radius: 4px;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  color: #FFFFFF;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-2::after {
  box-shadow: 0 0 0 2px #009BD8;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
}

<label
  class="emotion-0"
>
  <input
    aria-invalid="false"
    class="emotion-1"
    type="checkbox"
  />
  <div
    class="emotion-2"
  />
  <div
    class="emotion-3"
  >
    Taylor Swift
  </div>
</label>
`;

exports[`Checkbox renders snapshot of inverted 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 0;
  height: 0;
  margin: 0;
  padding: 0;
  overflow: hidden;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 18px;
  width: 18px;
  border-radius: 4px;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  color: #05192D;
  background-color: #213147;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #FFFFFF;
}

<label
  class="emotion-0"
>
  <input
    aria-invalid="false"
    class="emotion-1"
    type="checkbox"
  />
  <div
    class="emotion-2"
  />
  <div
    class="emotion-3"
  >
    Taylor Swift
  </div>
</label>
`;

exports[`Checkbox renders snapshot with error 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 0;
  height: 0;
  margin: 0;
  padding: 0;
  overflow: hidden;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 18px;
  width: 18px;
  border-radius: 4px;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  color: #FFFFFF;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
  border-color: #DD3400;
  box-shadow: 0 0 0 1px #DD3400;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
}

<label
  class="emotion-0"
>
  <input
    aria-invalid="true"
    class="emotion-1"
    type="checkbox"
  />
  <div
    class="emotion-2"
  />
  <div
    class="emotion-3"
  >
    Taylor Swift
  </div>
</label>
`;
